### ISSUE TITLE
Htu

### ISSUE THREAD

**Issue by gk-maker0105** [gk-maker0105/practicse#4]
I'll search for the detailed history of your cloud-related work specifically.

Now I have comprehensive context. Here's the detailed chronological history of your **cloud-related work**:

---

# ☁️ COMPLETE CLOUD PROJECT TIMELINE & FEATURE PROGRESS

---

### Datasets Acquired

| Dataset | Source | Images | Classes | Resolution | Special |
|---------|--------|--------|---------|------------|---------|
| **MGCD** | Google Drive | 8,000 | 7 | 1024×1024 fisheye | Weather data (temp, humidity, pressure, wind) |
| **CCSN** | Kaggle | 2,543 | 11 | 400×400 / 256×256 | Fine-grained cloud types |
| **TJNU-GCD** | Public URL | 19,000 | 7 | 512×512 | Largest, pre-split train/test |

### Class Mappings Created

```python
# 11-class CCSN → 7-class unified
CCSN_TO_UNIFIED = {
    'Cu': 'cumulus',
    'Ac': 'altocumulus', 
    'Ci': 'cirrus', 'Cs': 'cirrus', 'Ct': 'cirrus',  # merge high clouds
    'St': 'stratus', 'As': 'stratus', 'Ns': 'stratus',  # merge layer clouds
    'Sc': 'stratocumulus',
    'Cb': 'cumulonimbus',
    'Cc': 'mixed',
}

# TJNU-GCD already has 7 classes
TJNU_TO_UNIFIED = {
    '1_cumulus': 'cumulus', '2_altocumulus': 'altocumulus', 
    '3_cirrus': 'cirrus', '4_clearsky': 'clearsky', 
    '5_stratocumulus': 'stratocumulus', '6_cumulonimbus': 'cumulonimbus', 
    '7_mixed': 'mixed',
}
```

---

### Features Implemented (1,820 dimensions total)

| Feature Group | Count | Method | Purpose |
|---------------|-------|--------|---------|
| **HOG** | 1,764 | 9 orientations, 16×16 cells, 2×2 blocks | Edge/shape patterns |
| **Color Statistics** | 15 | Mean, std, skew, kurtosis, P25 per RGB | Color distribution |
| **Color Ratios** | 5 | R-G, R-B, G-B, R/B, c15=(B-R)/(B+R) | Spectral indices |
| **GLCM Texture** | 5 | Contrast, dissimilarity, homogeneity, energy, correlation | Texture patterns |
| **LBP Histogram** | 26 | 24 points, radius 3, uniform | Local texture |
| **Spatial** | 5 | Quadrant means (4) + edge density | Spatial distribution |

### Feature Extraction Code (Working)

```python
def extract_enhanced_features(img_path):
    img = cv2.imread(img_path)
    img = cv2.resize(img, (256, 256))
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    
    features = []
    
    # HOG (1764)
    hog_feat = hog(gray, orientations=9, pixels_per_cell=(16, 16), 
                   cells_per_block=(2, 2), feature_vector=True)
    features.extend(hog_feat)
    
    # Color stats (15)
    for c in range(3):
        channel = img[:,:,c].flatten()
        features.extend([np.mean(channel), np.std(channel), 
                        skew(channel), kurtosis(channel), np.percentile(channel, 25)])
    
    # Color ratios (5)
    r, g, b = img[:,:,0].mean(), img[:,:,1].mean(), img[:,:,2].mean()
    features.extend([r-g, r-b, g-b, r/(b+1e-6), (b-r)/(b+r+1e-6)])
    
    # GLCM (5)
    glcm = graycomatrix(gray, [1], [0], 256, symmetric=True)
    features.extend([graycoprops(glcm, 'contrast')[0,0], ...])
    
    # LBP (26)
    lbp = local_binary_pattern(gray, 24, 3, 'uniform')
    hist, _ = np.histogram(lbp.ravel(), bins=26, range=(0, 26))
    features.extend((hist / hist.sum()).tolist())
    
    # Spatial (5)
    h, w = gray.shape
    features.extend([gray[:h//2,:w//2].mean(), gray[:h//2,w//2:].mean(), 
                    gray[h//2:,:w//2].mean(), gray[h//2:,w//2:].mean()])
    features.append((np.abs(np.diff(gray,axis=1)).mean() + 
                    np.abs(np.diff(gray,axis=0)).mean())/2)
    
    return np.array(features)
```

---

### Baseline Accuracies (Verified)

| Dataset | Model | Train Acc | Test Acc | Features |
|---------|-------|-----------|----------|----------|
| **MGCD** | Random Forest | 100% | **60.0%** | 1,820 |
| **MGCD + Weather** | Random Forest | 100% | **57.3%** | 1,824 |
| **CCSN** | Gradient Boosting | — | **48.1%** | 1,820 |
| **TJNU-GCD** | Gradient Boosting | — | **54.5%** | 1,820 |

### ⚠️ Critical Finding: Original 96.71% Claim Was Wrong

The originally reported 96.71% accuracy was **incorrect** for 7-class classification:
- True holdout test accuracy: **55-61%**
- 100% train vs 61% test → **severe overfitting**
- Weather data actually **hurt** performance (-2.7%)

---

### Zero-Shot Cross-Dataset Transfer (All Failed ~15-20%)

| Train On | Test On | Accuracy | Drop |
|----------|---------|----------|------|
| MGCD | CCSN | **13.8%** | -75% |
| MGCD | TJNU | **20.3%** | -66% |
| CCSN | MGCD | **20.5%** | -57% |
| CCSN | TJNU | **16.7%** | -65% |
| TJNU | MGCD | **18.7%** | -66% |
| TJNU | CCSN | **20.9%** | -56% |

### Few-Shot Calibration (Success!)

| Transfer | Without Calibration | With 10 images | Improvement |
|----------|---------------------|----------------|-------------|
| Almeria → HYTA | 65.52% | **94.74%** | +29% |

**Key Insight:** 15 diverse features + 10 calibration images = successful transfer

---

### PCA Results

| Components | Cumulative Variance | Performance |
|------------|---------------------|-------------|
| 1 | 38.2% | — |
| 2 | 59.6% | — |
| 5 | 83.7% | — |
| 15 | **95.2%** | ✅ Recommended |
| 50 | 99%+ | — |

**Top PC1 Contributors:** Blue channel mean (0.42), Brightness (0.38), HSV Value (0.35)

**Top PC2 Contributors:** LBP texture variance (0.51), Edge density (0.44), Contrast (0.32)

### PLS Discriminant Analysis

| Feature | PLS Weight | Importance |
|---------|------------|------------|
| LBP texture variance | 0.312 | 1 ⭐ |
| Blue-Red ratio (c15) | 0.287 | 2 ⭐ |
| Edge density | 0.245 | 3 ⭐ |
| HSV Saturation | 0.198 | 4 |
| Brightness std | 0.176 | 5 |

**Key Insight:** Texture > Color for cloud classification

---

### Datasets Acquired

| Dataset | Images | Train/Val/Test | Resolution | Source |
|---------|--------|----------------|------------|--------|
| **SWIMSEG-2** | 1,013 | 861/101/51 | 600×600 | Singapore |
| **HYTA** | 32 | — | Variable (19 sizes!) | Multiple cameras |
| **Almería** | 818 | 770/—/48 | 512×512 | Spain (DLR) |
| **Total** | **1,863** | | | |

### Ground Truth Conventions (Critical!)

| Dataset | Convention | Fix Required |
|---------|------------|--------------|
| **SWIMSEG-2** | ⚠️ **INVERTED** (0=cloud, 255=sky) | Invert GT |
| **HYTA 2GT** | ❌ JPEG compressed, noisy | Use 3GT instead |
| **HYTA 3GT** | ✅ Clean (0=sky, 126=thin, 255=thick) | Threshold at 126 |
| **Almería** | Multi-class (0=ignore, 1=sky, 2-4=clouds) | Binary: ≥2 = cloud |

---

### HYTA Dataset (32 images)

| Model | F-score | IoU | Precision | Recall | Train Time |
|-------|---------|-----|-----------|--------|------------|
| **Coverage-Adaptive Ensemble** | **0.8326** | **0.7132** | 0.8471 | 0.8187 | — |
| PLS Regression (c15) | 0.8197 | 0.6945 | 0.8073 | 0.8325 | 0.03s |
| Naive Bayes (c15) | 0.8169 | 0.6900 | 0.8668 | 0.7725 | 0.02s |
| Random Forest (c15) | 0.7413 | 0.5889 | 0.6961 | 0.7929 | 21.06s |

### Coverage-Adaptive Strategy (Key Discovery!)

```
Coverage-based model selection:
├── Sparse clouds (<25%):  PLS wins (F=0.59 vs RF=0.31)
├── Mid-range (25-50%):    RF wins (F=0.55 vs PLS=0.41)
└── Heavy clouds (>50%):   PLS wins (F=0.75 vs RF=0.73)
```

### Almeria Dataset (770 train, 48 test)

| Task | Best Model | Accuracy | F1-Score |
|------|------------|----------|----------|
| **Binary** (Cloud/Clear) | Gradient Boosting | **98.05%** | 0.9806 |
| **4-Class** (Clear/Low/Mid/High) | MLP | **92.21%** | 0.8508 |

**Top 5 Features:**
1. cloud_fraction (18.63%)
2. H_mean (9.69%)
3. GB_diff (4.71%)
4. glcm_energy (4.56%)
5. RB_ratio (4.51%)

---

### U-Net Training (SWIMSEG-2)

| Configuration | Test IoU | Test F1 |
|---------------|----------|---------|
| U-Net + MobileNetV2 (threshold 0.5) | 48.1% | — |
| U-Net + Color Prior | — | — |
| **Ensemble (Union)** | **60.1%** | — |

### Cross-Dataset Generalization

| Train On | Test On | IoU |
|----------|---------|-----|
| SWIMSEG | SWIMSEG Test | 60.1% |
| SWIMSEG | HYTA | **22.6%** (61% drop!) |

**Key Finding:** Ensemble provides +42% relative improvement even on unseen HYTA data

---

### Combined Binary Segmentation Dataset

| Source | Train | Val | Test | Notes |
|--------|-------|-----|------|-------|
| HYTA | 0 | 0 | 32 | Test only (different cameras) |
| ALMERIA | 616 | — | 48 | Convert to binary |
| SWIMSEG | 861 | 101 | 51 | Invert GT |
| **Total** | **1,477** | **101** | **131** | |

### Unified Dataset JSON Manifest Created

```json
{
  "train": {"images": [...], "masks": [...], "sources": [...]},
  "val": {"images": [...], "masks": [...], "sources": [...]},
  "test": {"images": [...], "masks": [...], "sources": [...]}
}
```

---

### Image Statistics Analyzed

| Metric | HYTA | ALMERIA | SWIMSEG |
|--------|------|---------|---------|
| **Brightness** | 123 | **41** (3x darker!) | 148 |
| **R/B Ratio** | 0.68 | **0.99** (gray) | 0.71 |
| **Cloud Regions** | 1-2 | **24 median** | 1-2 |
| **Blue Dominance** | 0.35 | 0.34 | 0.35 |

### ⚠️ Domain Shift Discovery

```
UMAP Embedding Distances:
├── HYTA ↔ SWIMSEG:   0.59  (nearly identical)
└── ALMERIA ↔ both:   15.17 (completely different domain!)
```

### Cleanlab Analysis

- **185 potential label issues** (37% of sample)
- Primarily in ALMERIA multi-class → binary conversion

---

## 📅 **PHASE 11: Pending/Incomplete Work**

| Task | Status | Notes |
|------|--------|-------|
| ALMERIA Part 2 EDA | 🔄 Pending | ~385 images remaining |
| SWIMSEG Parts 1-3 EDA | 🔄 Pending | ~1,013 images remaining |
| Domain adaptation training | 📋 Planned | ALMERIA needs special handling |
| Multi-dataset U-Net training | 📋 Planned | With domain-specific normalization |

---

### ✅ Verified & Working

| Category | Achievement |
|----------|-------------|
| **Feature Extraction** | 1,820-dim features (HOG+LBP+Color+GLCM+Spatial) |
| **Classification Baseline** | 50-60% for 7-class (verified) |
| **Segmentation Baseline** | 60% IoU with ensemble on SWIMSEG |
| **Coverage-Adaptive** | PLS/RF switching based on cloud density |
| **Transfer Success** | 10 calibration images → +29% improvement |
| **Transfer Failure** | Zero-shot cross-dataset drops 60-75% |

### ❌ Disproven

| Claim | Reality |
|-------|---------|
| 96.71% MGCD accuracy | Actually **60%** (was overfitting or different task) |
| Weather data helps +4.14% | Actually **-2.7%** (hurt performance) |

### 🔮 Key Insights for Future

1. **Texture > Color** for cloud classification (PLS analysis)
2. **15 features + 10 calibration images** = successful transfer
3. **ALMERIA is different domain** (needs brightness normalization or separate training)
4. **Coverage-adaptive ensemble** beats single models
5. **Classical ML caps at ~60%** for 7-class; deep learning needed for 85%+

---

Would you like me to:
1. **Resume the pending EDA** on SWIMSEG/ALMERIA?
2. **Create a complete guide document** consolidating all this?
3. **Design the domain adaptation training pipeline**?
